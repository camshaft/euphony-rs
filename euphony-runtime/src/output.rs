use euphony_sc::project::Project;
use std::{io, sync::Arc};

pub type Handle = Arc<dyn Output>;

pub trait Output: 'static + Project + Send + Sync {
    fn finish(&self) -> io::Result<()>;
}
