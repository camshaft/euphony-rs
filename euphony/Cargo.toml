[package]
name = "euphony"
description = "music composition environment for programmers"
version = "0.1.1"
authors = ["Cameron Bytheway <bytheway.cameron@gmail.com>"]
edition = "2021"
license = "MIT"

[dependencies]
bach = "=0.0.6"
euphony-buffer = { version = "0.1", path = "../euphony-buffer", default-features = false }
euphony-command = { version = "0.1", path = "../euphony-command" }
euphony-macros = { version = "0.2", path = "../euphony-macros" }
euphony-samples = { version = "0.1", path = "../euphony-samples" }
euphony-units = { version = "0.1", path = "../euphony-units" }
futures = "0.3"
midly = { version = "0.5", default-features = false, features = ["std"] }
once_cell = "1"
pin-project = "1"
rand = { version = "0.8", default-features = false }

[dev-dependencies]
insta = "1"
euphony-node = { version = "0.1", path = "../euphony-node", features = ["reflect"] }
