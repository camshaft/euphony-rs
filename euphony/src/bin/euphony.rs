fn main() {
    euphony_cli::main()
}
