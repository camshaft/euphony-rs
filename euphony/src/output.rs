pub use euphony_command::message;
use once_cell::sync::Lazy;
use std::{fs, io, path::Path, sync::Mutex, time::Duration};

use message::Message;

pub trait Output: 'static + Send + Sync {
    fn emit(&mut self, message: Message);
}

pub struct Stdout;

impl Output for Stdout {
    fn emit(&mut self, message: Message) {
        println!("{}", message);
    }
}

pub struct File(io::BufWriter<fs::File>);

impl File {
    pub fn create(path: &Path) -> io::Result<Self> {
        let file = fs::File::create(path)?;
        let file = io::BufWriter::new(file);
        Ok(Self(file))
    }
}

impl Output for File {
    fn emit(&mut self, message: Message) {
        message.write(&mut self.0).unwrap();
    }
}

static OUTPUT: Lazy<Mutex<Box<dyn Output>>> = Lazy::new(|| Mutex::new(Box::new(Stdout)));

pub fn set_output(output: Box<dyn Output>) {
    *OUTPUT.lock().unwrap() = output;
}

pub fn set_file(path: &Path) {
    let file = File::create(path);
    let file = file.expect("could not create output file");
    let output = Box::new(file);
    set_output(output)
}

pub(crate) fn emit(message: Message) {
    OUTPUT.lock().unwrap().emit(message);
}

pub(crate) fn set_seed(seed: u64) {
    emit(Message::SetSeed { seed });
}

pub(crate) fn set_group_name(id: u64, name: &str) {
    emit(Message::SetGroupName { id, name });
}

pub(crate) fn advance(amount: Duration) {
    emit(Message::AdvanceTime { amount })
}

pub(crate) fn finish() {
    emit(Message::Finish);
}
