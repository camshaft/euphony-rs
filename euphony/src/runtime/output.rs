use euphony_command::message::{self, Message};
use once_cell::sync::Lazy;
use std::{sync::Mutex, time::Duration};

pub trait Output: 'static + Send + Sync {
    fn emit(&mut self, message: Message);
}

struct Stdout;

impl Output for Stdout {
    fn emit(&mut self, message: Message) {
        println!("{}", message);
    }
}

static OUTPUT: Lazy<Mutex<Box<dyn Output>>> = Lazy::new(|| Mutex::new(Box::new(Stdout)));

pub fn set_output(output: Box<dyn Output>) {
    *OUTPUT.lock().unwrap() = output;
}

pub fn emit(message: Message) {
    OUTPUT.lock().unwrap().emit(message);
}

pub fn set_seed(seed: u64) {
    emit(Message::SetSeed { seed });
}

pub fn set_group_name(id: u64, name: &str) {
    emit(Message::SetGroupName { id, name });
}

pub fn advance(amount: Duration) {
    emit(Message::AdvanceTime { amount })
}

pub fn finish() {
    emit(Message::Finish);
}
