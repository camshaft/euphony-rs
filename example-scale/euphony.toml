[synths]
chiplead = "gh://sonic-pi-net/sonic-pi/etc/synthdefs/compiled/sonic-pi-chiplead.scsyndef"
bass = "gh://sonic-pi-net/sonic-pi/etc/synthdefs/compiled/sonic-pi-blade.scsyndef"

[samples]
bd = "clean://tidalcycles/sounds-tr808-fischer/bd8"
sd = "clean://tidalcycles/sounds-tr808-fischer/sd8"
hh = "clean://tidalcycles/sounds-tr808-fischer/ch8"
cy = "clean://tidalcycles/sounds-tr808-fischer/cy8"
ht = "clean://tidalcycles/sounds-tr808-fischer/ht8"
lt = "clean://tidalcycles/sounds-tr808-fischer/lt8"
mt = "clean://tidalcycles/sounds-tr808-fischer/mt8"
