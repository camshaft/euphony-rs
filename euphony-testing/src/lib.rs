use core::panic::Location;
use euphony::output::{self, message::Message};
use std::{
    future::Future,
    sync::{Arc, Mutex},
};

#[derive(Clone, Default)]
struct List(Arc<Mutex<String>>);

impl output::Output for List {
    fn emit(&mut self, message: Message) {
        let mut out = self.0.lock().unwrap();
        out.push_str(&format!("{}\n", message));
    }
}

#[track_caller]
pub fn start<F>(f: F)
where
    F: 'static + Future<Output = ()> + Send,
{
    let list = List::default();
    output::set_output(Box::new(list.clone()));

    euphony::runtime::Runtime::new(0).block_on(f);

    let result = list.0.lock().unwrap();

    let name = Location::caller().file();
    let name = name.split('/').last().unwrap();
    let name = name.trim_end_matches(".rs");
    insta::assert_display_snapshot!(name, *result);
}
