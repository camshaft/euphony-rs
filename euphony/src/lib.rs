pub use euphony_core::*;
pub use euphony_macros::*;
pub use euphony_sc;

mod args;
pub mod ext;
pub mod runtime;

#[macro_export]
macro_rules! prelude {
    () => {
        #[cfg(euphony_assets)]
        include!(concat!(env!("OUT_DIR"), "/euphony_assets.rs"));

        use euphony::prelude::*;
    };
}

pub mod prelude {
    pub use crate::{
        ext::*,
        pitch::Interval,
        runtime::{output::track, spawn, time::scheduler},
        time::{Beat, Tempo},
    };
    pub use euphony_sc::{self, params, synthdef, track::Track};
}
