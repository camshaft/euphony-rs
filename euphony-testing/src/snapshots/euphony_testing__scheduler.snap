---
source: euphony-testing/src/lib.rs
expression: "*result"
---
{"SetSeed":{"seed":0}}
{"SynthDef":{"id":0,"definition":{}}}
{"SetGroupName":{"id":0,"name":"main"}}
{"Spawn":{"id":0,"synthdef":0,"group":0}}
{"Set":{"id":0,"parameter":0,"value":0.0}}
{"SynthDef":{"id":1,"definition":{}}}
{"Spawn":{"id":1,"synthdef":1,"group":0}}
{"Set":{"id":1,"parameter":0,"value":0.0}}
{"AdvanceTime":{"amount":{"secs":0,"nanos":125000000}}}
{"Drop":{"id":1}}
{"Spawn":{"id":2,"synthdef":1,"group":0}}
{"Set":{"id":2,"parameter":1,"value":0.25}}
{"AdvanceTime":{"amount":{"secs":0,"nanos":125000000}}}
{"Drop":{"id":0}}
{"Spawn":{"id":3,"synthdef":0,"group":0}}
{"Set":{"id":3,"parameter":1,"value":0.5}}
{"Drop":{"id":2}}
{"Spawn":{"id":4,"synthdef":1,"group":0}}
{"Set":{"id":4,"parameter":2,"value":0.5}}
{"AdvanceTime":{"amount":{"secs":0,"nanos":125000000}}}
{"Drop":{"id":4}}
{"Spawn":{"id":5,"synthdef":1,"group":0}}
{"Set":{"id":5,"parameter":3,"value":0.75}}
{"AdvanceTime":{"amount":{"secs":0,"nanos":125000000}}}
{"Drop":{"id":3}}
{"Spawn":{"id":6,"synthdef":0,"group":0}}
{"Set":{"id":6,"parameter":2,"value":1.0}}
{"Drop":{"id":5}}
{"Spawn":{"id":7,"synthdef":1,"group":0}}
{"Set":{"id":7,"parameter":4,"value":1.0}}
{"AdvanceTime":{"amount":{"secs":0,"nanos":125000000}}}
{"Drop":{"id":7}}
{"Spawn":{"id":8,"synthdef":1,"group":0}}
{"Set":{"id":8,"parameter":5,"value":1.25}}
{"AdvanceTime":{"amount":{"secs":0,"nanos":125000000}}}
{"Drop":{"id":6}}
{"Spawn":{"id":9,"synthdef":0,"group":0}}
{"Set":{"id":9,"parameter":3,"value":1.5}}
{"Drop":{"id":8}}
{"Spawn":{"id":10,"synthdef":1,"group":0}}
{"Set":{"id":10,"parameter":6,"value":1.5}}
{"AdvanceTime":{"amount":{"secs":0,"nanos":125000000}}}
{"Drop":{"id":10}}
{"Spawn":{"id":11,"synthdef":1,"group":0}}
{"Set":{"id":11,"parameter":7,"value":1.75}}
{"AdvanceTime":{"amount":{"secs":0,"nanos":125000000}}}
{"Drop":{"id":9}}
{"Spawn":{"id":12,"synthdef":0,"group":0}}
{"Set":{"id":12,"parameter":4,"value":2.0}}
{"Drop":{"id":11}}
{"Spawn":{"id":13,"synthdef":1,"group":0}}
{"Set":{"id":13,"parameter":8,"value":2.0}}
{"AdvanceTime":{"amount":{"secs":0,"nanos":125000000}}}
{"Drop":{"id":13}}
{"Spawn":{"id":14,"synthdef":1,"group":0}}
{"Set":{"id":14,"parameter":9,"value":2.25}}
{"AdvanceTime":{"amount":{"secs":0,"nanos":125000000}}}
{"Drop":{"id":12}}
{"Spawn":{"id":15,"synthdef":0,"group":0}}
{"Set":{"id":15,"parameter":5,"value":2.5}}
{"Drop":{"id":14}}
{"AdvanceTime":{"amount":{"secs":0,"nanos":250000000}}}
{"Drop":{"id":15}}
{"Spawn":{"id":16,"synthdef":0,"group":0}}
{"Set":{"id":16,"parameter":6,"value":3.0}}
{"AdvanceTime":{"amount":{"secs":0,"nanos":250000000}}}
{"Drop":{"id":16}}
{"Spawn":{"id":17,"synthdef":0,"group":0}}
{"Set":{"id":17,"parameter":7,"value":3.5}}
{"AdvanceTime":{"amount":{"secs":0,"nanos":250000000}}}
{"Drop":{"id":17}}
{"Spawn":{"id":18,"synthdef":0,"group":0}}
{"Set":{"id":18,"parameter":8,"value":4.0}}
{"AdvanceTime":{"amount":{"secs":0,"nanos":250000000}}}
{"Drop":{"id":18}}
{"Spawn":{"id":19,"synthdef":0,"group":0}}
{"Set":{"id":19,"parameter":9,"value":4.5}}
{"AdvanceTime":{"amount":{"secs":0,"nanos":250000000}}}
{"Drop":{"id":19}}
"Finish"

